src/extensions/geo-export/export.ts

```diff
@@ -0,0 +1,319 @@
+
+        const aPosition = values.aPosition.ref.value;
+        const aGroup = values.aGroup.ref.value;
+        const instanceCount = values.instanceCount.ref.value;
+        const vertexCount = values.uVertexCount.ref.value;
+
+        for (let instanceIndex = 0; instanceIndex < instanceCount; ++instanceIndex) {
+            const state = MeshBuilder.createState(512, 256);
+
+            for (let i = 0; i < vertexCount; i += 4) {
+                v3fromArray(center, aPosition, i * 3);
+
+                const group = aGroup[i];
+                const radius = ObjExporter.getSize(values, instanceIndex, group);
+                state.currentGroup = group;
+                addSphere(state, center, radius, 2);
+            }
+
+            const mesh = MeshBuilder.getMesh(state);
+            const vertices = mesh.vertexBuffer.ref.value;
+            const indices = mesh.indexBuffer.ref.value;
+            const normals = mesh.normalBuffer.ref.value;
+            const groups = mesh.groupBuffer.ref.value;
+            this.addMeshWithColors(vertices, indices, normals, groups, vertices.length / 3, indices.length, values, instanceIndex);
+        }
+    }
+
+    private addCylinders(values: CylindersValues) {
+        const start = Vec3();
+        const end = Vec3();
+
+        const aStart = values.aStart.ref.value;
+        const aEnd = values.aEnd.ref.value;
+        const aScale = values.aScale.ref.value;
+        const aCap = values.aCap.ref.value;
+        const aGroup = values.aGroup.ref.value;
+        const instanceCount = values.instanceCount.ref.value;
+        const vertexCount = values.uVertexCount.ref.value;
+
+        for (let instanceIndex = 0; instanceIndex < instanceCount; ++instanceIndex) {
+            const state = MeshBuilder.createState(512, 256);
+
+            for (let i = 0; i < vertexCount; i += 6) {
+                v3fromArray(start, aStart, i * 3);
+                v3fromArray(end, aEnd, i * 3);
+
+                const group = aGroup[i];
+                const radius = ObjExporter.getSize(values, instanceIndex, group) * aScale[i];
+                const cap = aCap[i];
+                const topCap = cap === 1 || cap === 3;
+                const bottomCap = cap >= 2;
+                const cylinderProps = { radiusTop: radius, radiusBottom: radius, topCap, bottomCap, radialSegments: 32 };
+                state.currentGroup = aGroup[i];
+                addCylinder(state, start, end, 1, cylinderProps);
+            }
+
+            const mesh = MeshBuilder.getMesh(state);
+            const vertices = mesh.vertexBuffer.ref.value;
+            const indices = mesh.indexBuffer.ref.value;
+            const normals = mesh.normalBuffer.ref.value;
+            const groups = mesh.groupBuffer.ref.value;
+            this.addMeshWithColors(vertices, indices, normals, groups, vertices.length / 3, indices.length, values, instanceIndex);
+        }
+    }
+
+    add(renderObject: GraphicsRenderObject) {
```

Comment:
Do you think it would be worthwhile to add ``RuntimeContext`` reference here to get partial progress/cancellation for large meshes?

```ts
async add(renderObject: GraphicsRenderObject, ctx: RuntimeContext) {
  ...
  // pass the ctx to addMesh and other functions and do this every N elements (for large enough N)
  if (i % N === 0 && ctx.shouldUpdate) {
    await ctx.update(...);
  }
  ...
}
```

Reply:
Hi David. Thank you for the feedback. Yes, it's definitely worthwhile. I have made these functions asynchronous. Users can now cancel the exporting for large meshes.

I'm not sure what's the best way to show partial progress though. I can do something like
```js
for (let i = 0; i < drawCount; i += 3) {
    if (i % 3000 === 0 && ctx.shouldUpdate) await ctx.update({ message: 'Rendering...', current: i, max: drawCount });
    ...
}
```
But then the number is reset for every RenderObject. Another approach that I can think of is to compute beforehand the total number of faces of all RenderObjects. But it seems complicated.

Reply:
I can see two more options:
- Show show the current render object in the message. E.g. ``update({ message: `Exporting object ${i}/${objCount}`, current: ..., max: ... })``
- Return a ``Task`` by the ``add`` function and run it with ``runAsChild``. Since there won't event be that many objects to render, this should not have any negative perf. implications.
```ts
add(renderObject: GraphicsRenderObject, ctx: RuntimeContext) {
    return Task.create('Export Object', ctx => {
        if (!renderObject.state.visible) return;

        switch (renderObject.type) {
            case 'mesh':
                // can return the promise without using async function here instead of
                // async ctx => .... await add(..., ctx); break;
                return this.addMesh(renderObject.values as MeshValues, ctx);
            // ...
        }
    });
}

//////////////////////////

for (let i = 0, il = renderObjects.length; i < il; ++i) {
    await objExporter.add(renderObjects[i]).runAsChild(ctx);
    await ctx.update({ current: i });
}
```

Reply:
Thank you for your suggestions. I've updated the code to show partial progress.